docs(recipes): Column-based access (#3269)
cube-js/cube.js: docs/content/Examples-Tutorials-Recipes/Column-based-access.md

---
title: Column-Based Access
permalink: /recipes/column-based-access
category: Examples & Tutorials
subCategory: Access control
menuOrder: 2
---

## Use case

We want to manage user access to different data depending on a database
relationship. In the recipe below, we will manage supplier access to their
products. A supplier can't see other supplier's products.

## Data schema

To implement column-based access, we will use supplier's email from

> [JSON Web Token](https://cube.dev/docs/security), and the

Reviewer: ```suggestion
a [JSON Web Token](https://cube.dev/docs/security), and the
```

> [queryRewrite](https://cube.dev/docs/security/context#using-query-rewrite)

Reviewer: ```suggestion
[`queryRewrite`](https://cube.dev/docs/security/context#using-query-rewrite)
```

extension point to manage data access.

We have `Products` and `Suppliers` tables with `hasOne` relationship from
products to suppliers:

```javascript
// schema/Products.js

joins: {
    Suppliers: {
        sql: `${Suppliers}.id = ${CUBE}.supplier_id`,
        relationship: `hasOne`
    }

> }

Reviewer: It might be better to show both cubes fully defined in the example above, so as to not leave any ambiguity

Author: Added both cubes 👌

```

> Let's add the supplier email filter if a query exists `Products` dimension:

Reviewer: ```suggestion
Let's add the supplier email filter if a query includes any dimension from the `Products` cube:
```

```javascript
module.exports = {
  queryRewrite: (query, { securityContext }) => {
    const queryDimensions = Array.from(
      query.dimensions,
      (element) => element.split('.')[0]
    );

    if (queryDimensions.includes('Products')) {
      query.filters.push({
        member: `Suppliers.email`,
        operator: 'equals',
        values: [securityContext.email],
      });
    }

    return query;
  },
};
```

## Query

To get the supplier's products, we will send two identical requests with

> different emails at authorization token.

Reviewer: ```suggestion
different emails inside the JWT.
```

---
title: Column-Based Access
permalink: /recipes/column-based-access
category: Examples & Tutorials
subCategory: Access control
menuOrder: 2
---

## Use case

We want to manage user access to different data depending on a database
relationship. In the recipe below, we will manage supplier access to their
products. A supplier can't see other supplier's products.

## Data schema

To implement column-based access, we will use supplier's email from
a [JSON Web Token](https://cube.dev/docs/security), and the
[`queryRewrite`](https://cube.dev/docs/security/context#using-query-rewrite)
extension point to manage data access.

> We have `Products` and `Suppliers` tables with `hasOne` relationship from

Reviewer 2: ```suggestion
We have `Products` and `Suppliers` cubes with `hasOne` relationship from
```

products to suppliers:

```javascript

> // schema/Products.js
> 

Reviewer 2: ```suggestion
```

It's obvious from the code, right? 😄

cube(`Products`, {
  sql: `SELECT * FROM public.products`,

  joins: {
    Suppliers: {
      sql: `${Suppliers}.id = ${CUBE}.supplier_id`,
      relationship: `hasOne`
    }
  },
  
  dimensions: {
    name: {
      sql: `name`,
      type: `string`
    }
  }
});
```
```javascript

> // schema/Suppliers.js
> 

Reviewer 2: ```suggestion
```

Ditto.

cube(`Suppliers`, {
  sql: `SELECT * FROM public.suppliers`,

  dimensions: {
    email: {
      sql: `email`,
      type: `string`
    }
  }
});
```

## Configuration

> Let's add the supplier email filter if a query includes any dimension from the `Products` cube:

Reviewer 2: ```suggestion
Let's add the supplier email filter if a query includes any dimensions from the `Products` cube:
```

Reviewer 2: I wonder why only dimensions? 😄 (I personally think it's better to say "any dimensions or measures".)

```javascript
module.exports = {
  queryRewrite: (query, { securityContext }) => {

>     const queryDimensions = Array.from(
>       query.dimensions,
>       (element) => element.split('.')[0]
>     );

Reviewer 2: ```suggestion
    const cubeNames = [
      ...(Array.from(query.measures, (e) => e.split('.')[0])),
      ...(Array.from(query.dimensions, (e) => e.split('.')[0])),
    ];
```

WDYT? TBH, haven't tested this 😄

Author: It's a good suggestion, thanks!

    if (queryDimensions.includes('Products')) {
      query.filters.push({
        member: `Suppliers.email`,
        operator: 'equals',

>         values: [securityContext.email],

Reviewer 2: Also, I think it's great to check if the email is present, fail the query otherwise.

      });
    }

    return query;
  },
};
```

## Query

To get the supplier's products, we will send two identical requests with

> different emails inside the JWT.

Reviewer 2: ```suggestion
different emails inside JWTs.
```
